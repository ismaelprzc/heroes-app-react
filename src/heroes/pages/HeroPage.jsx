import { useMemo } from "react";
import { Navigate, useNavigate, useParams } from "react-router-dom";
import { getHeroeById } from "../helpers";

export const HeroPage = () => {
  const { id } = useParams();

  const hero = useMemo(() => getHeroeById(id), [id]);

  const navigate = useNavigate();

  const back = () => {
    navigate(-1);
  };

  /*  console.log(hero)
   */

  //si no existe el heroe por url
  if (!hero) {
    return <Navigate to="/marvel" />;
  }

  return (
    <div className="row mt-5 animate__animated animate__fadeInLeft">
      <h3>{hero.superhero}</h3>
      <div className="col-4">
        <img
          src={`/assets/${id}.jpg`}
          alt={hero.superhero}
          className="img-thumbnail animate__animated animate__fadeInLeft"
        />
      </div>

      <div className="col-8">
        <ul className="list-group list-group-flush">
          <li className="list-group-item">
            <b>Alter Ego: </b> {hero.alter_ego}
          </li>
          <li className="list-group-item">
            <b>Publisher: </b> {hero.publisher}
          </li>
          <li className="list-group-item">
            <b>First Appearance: </b> {hero.first_appearance}
          </li>
        </ul>
        <h5 className="mt-3">Characters</h5>
        <ul>
          <li>{hero.characters}</li>
        </ul>

        <button id="btn" className="btn btn-primary" onClick={back}>
          Volver
        </button>
      </div>
    </div>
  );
};
