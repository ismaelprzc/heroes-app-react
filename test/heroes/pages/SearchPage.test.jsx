const { render, screen, fireEvent } = require("@testing-library/react");
const { MemoryRouter } = require("react-router-dom");
import { SearchPage } from "../../../src/heroes/pages/SearchPage";

const mockedUseNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUseNavigate,
}));

describe("Pruebas en <SearchPage>", () => {
  test("debe de mostrarse correctamente con valores por defecto", () => {
    const { container } = render(
      <MemoryRouter>
        <SearchPage></SearchPage>
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });

  test("debe de mostrar a Batman y el input con el valor del querystring", () => {
    render(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <SearchPage></SearchPage>
      </MemoryRouter>
    );

    const inputValue = screen.getByRole("textbox");
    expect(inputValue.value).toBe("batman");

    const img = screen.getByRole("img");
    expect(img.src).toContain("/assets/dc-batman.jpg");

    const alert = screen.getByLabelText("alert-danger");
    expect(alert.style.display).toBe("none");
  });

  test("debe de mostrar un error si no se encuentra el heroe (batman123)", () => {
    render(
      <MemoryRouter initialEntries={["/search?q=batman123"]}>
        <SearchPage></SearchPage>
      </MemoryRouter>
    );
    const alert = screen.getByLabelText("alert-danger");
    expect(alert.style.display).toBe("");
  });

  test("debe de llamar el navigate a la pantalla nueva", () => {
    const inputValue = "superman";
    render(
      <MemoryRouter initialEntries={["/search?q=batman123"]}>
        <SearchPage></SearchPage>
      </MemoryRouter>
    );

    const input = screen.getByRole("textbox");
    fireEvent.change(input, {
      target: { name: "searchText", value: inputValue },
    });

    const form = screen.getByRole("form");
    fireEvent.submit(form);

    expect(mockedUseNavigate).toHaveBeenCalledWith(`?q=${inputValue}`);
  });
});
