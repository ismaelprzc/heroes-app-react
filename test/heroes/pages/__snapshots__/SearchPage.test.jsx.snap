// Vitest Snapshot v1

exports[`Pruebas en <SearchPage> > debe de mostrar a Batman y el input con el valor del querystring 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
       
      <form>
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-primary mt-1"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        Search a hero!
      </div>
      <div
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        No hero with 
        <b />
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchPage> > debe de mostrar a Batman y el input con el valor del querystring correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
       
      <form>
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-primary mt-1"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        Search a hero!
      </div>
      <div
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        No hero with 
        <b />
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchPage> > debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
       
      <form>
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-primary mt-1"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        Search a hero!
      </div>
      <div
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        No hero with 
        <b />
      </div>
    </div>
  </div>
</div>
`;
